api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java

```diff
@@ -0,0 +1,187 @@
+		
+		Encounter encounter2 = context.getAttachmentEncounter(patient, visit, provider);
+		Encounter encounter3 = context.getAttachmentEncounter(patient, visit, provider);
+		
+		// adding a non-complex obs from encounter1
+		Obs otherObs1 = new Obs();
+		otherObs1.setConcept(cs.getConcept(3));
+		otherObs1.setObsDatetime(new Date());
+		otherObs1.setEncounter(encounter1);
+		otherObs1.setPerson(patient);
+		otherObs1.setValueText("Some text value for a test obs.");
+		otherObs1 = os.saveObs(otherObs1, null);
+		
+		// adding a non-complex obs from encounter2
+		Obs otherObs2 = new Obs();
+		otherObs2.setConcept(cs.getConcept(3));
+		otherObs2.setObsDatetime(new Date());
+		otherObs2.setEncounter(encounter2);
+		otherObs2.setPerson(patient);
+		otherObs2.setValueText("Some text value for a test obs.");
+		otherObs2 = os.saveObs(otherObs2, null);
+		
+		List<Attachment> expectedAttachments = testHelper.saveComplexObs(encounter1, 1, 1);
+		expectedAttachments.addAll(testHelper.saveComplexObs(encounter2, 2, 2));
+		expectedAttachments.addAll(testHelper.saveComplexObs(encounter3, 3, 3));
+		
+		// replay
+		List<Attachment> actualAttachments = as.getAttachments(patient, visit, true);
+		
+		// verify
+		Assert.assertArrayEquals(
+		    expectedAttachments.stream().map(Attachment::getUuid).collect(Collectors.toList()).toArray(),
+		    actualAttachments.stream().map(Attachment::getUuid).collect(Collectors.toList()).toArray());
+	}
+	
+	@Test
+	public void getAttachments_shouldReturnIsolatedAttachments() throws Exception {
+		
+		// setup
+		Encounter encounter = testHelper.getTestEncounter();
+		List<Attachment> attachments = testHelper.saveComplexObs(encounter, 1, 1);
+		Patient patient = ps.getPatient(2);
+		List<Attachment> expectedAttachments = new ArrayList<>();
+		
+		// Saves a complex obs as if they had been saved relevant to the attachments not associated with any visits or encounters.
+		for (int i = 0; i < 2; i++) {
+			Obs obs = new Obs();
+			ConceptComplex conceptComplex = context.getConceptService()
+			        .getConceptComplex(attachments.get(0).getObs().getConcept().getConceptId());
+			obs.setConcept(conceptComplex);
+			byte[] randomData = new byte[20];
+			new Random().nextBytes(randomData);
+			MockMultipartFile multipartRandomFile = new MockMultipartFile("1", "1", "application/octet-stream", randomData);
+			obs.setComplexData(
+			    context.complexDataHelper.build(ValueComplex.INSTRUCTIONS_DEFAULT, multipartRandomFile.getOriginalFilename(),
+			        multipartRandomFile.getBytes(), multipartRandomFile.getContentType()).asComplexData());
+			obs.setObsDatetime(new Date());
+			obs.setPerson(patient);
+			obs.setValueText("Some text value for a test obs.");
+			obs = os.saveObs(obs, null);
+			expectedAttachments.add(new Attachment(obs));
+		}
+		
+		// Saves a complex obs as if they had been saved outside to the attachments not associated with any visits or encounters.
+		for (int j = 0; j < 2; j++) {
```

Comment:
Reuse `i` as a name for the counter variable. Those loop counters' scopes are _inside_ the `for` statement, you can have as many successive `for` statements that keep reusing `i` as the counter variable name.

Reply:
If you have done this in other places, please apply the same logic there as well.